presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SimplifyExpressions.java

```diff
@@ -285,28 +277,21 @@ public Expression rewriteLogicalBinaryExpression(LogicalBinaryExpression node, N
         }
     }
 
-    private static class GroupingOperationContext
+    private static class GroupingOperationToConstant
+            extends ExpressionRewriter<Void>
     {
-        private final boolean isGroupIdNodePresent;
+        private final PlanNode source;
 
-        GroupingOperationContext(boolean isGroupIdNodePresent)
+        GroupingOperationToConstant(PlanNode source)
         {
-            this.isGroupIdNodePresent = isGroupIdNodePresent;
+            this.source = source;
         }
 
-        public boolean getIsGroupIdNodePresent()
-        {
-            return isGroupIdNodePresent;
-        }
-    }
-
-    private static class GroupingOperationToConstant
-            extends ExpressionRewriter<GroupingOperationContext>
-    {
         @Override
-        public Expression rewriteFunctionCall(FunctionCall node, GroupingOperationContext context, ExpressionTreeRewriter<GroupingOperationContext> treeRewriter)
+        public Expression rewriteFunctionCall(FunctionCall node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
         {
-            if (!context.getIsGroupIdNodePresent() && node.getName().toString().equals("grouping")) {
+            // A grouping() is preceded by a GroupId node when the source node is projecting a groupid symbol
+            if (!source.getOutputSymbols().contains(new Symbol("groupid")) && node.getName().toString().equals(GROUPING)) {
```

Comment:
Generally assuming symbol name is a bad idea (plan structure might be changed by other optimizers, symbols can be renamed/aliased, there can be multiple symbols with same prefix, `groupid`, `groupid_1`). I would prefer that you extract exact symbol instance from an appropriate `PlanNode`.